src/mo_frontend/typing.ml

```diff
@@ -1196,6 +1209,61 @@ let error_bin_op env at t1 t2 =
     display_typ_expand t1
     display_typ_expand t2
 
+let compare_pat_field (pf1 : pat_field) (pf2 : pat_field) =
+  compare pf1.it.id.it pf2.it.id.it
+
+let rec combine_pat_fields_srcs env t tfs (pfs : pat_field list) : unit =
+  match tfs, pfs with
+  | _, [] | [], _ -> ()
+  | T.{lab; typ = Typ _; _}::tfs', _ ->  (* TODO: remove the namespace hack *)
+    combine_pat_fields_srcs env t tfs' pfs
+  | T.{lab; typ; src}::tfs', pf::pfs' ->
+    match compare pf.it.id.it lab with
+    | -1 -> combine_pat_fields_srcs env t [] pfs
+    | +1 -> combine_pat_fields_srcs env t tfs' pfs
+    | _ ->
+      combine_pat_srcs env typ pf.it.pat;
+      combine_pat_fields_srcs env t tfs' pfs';
+
+and combine_id_srcs env t id : unit =
+  match T.Env.find_opt id.it env.vals with
+  | None -> ()
+  | Some (t', _, _, _) ->
+    (* Use [sub] to merge the fields of sources, if one type is indeed a subtype
```

Comment:
Even if they aren't subtype, the merging will  persist because of the destructive update, will it not?